cam6_4_120: Complete CCPPization of HB PBL scheme and vertical diffusion solver (non-WACCM)

jimmielin opened the pull request:
Companion PR: https://github.com/ESCOMP/atmospheric_physics/pull/270

Purpose of changes (include the issue number and title text for each relevant GitHub issue):
- Completes #1300 by CCPPizing compute_hb_diff, compute_hb_free_atm_diff, pblintd, trbintd
- Cleanup, refactoring, and optimization of HB
- Partially completes #1205; and completes #1314 by CCPPizing partial interstitial logic of vertical_diffusion_tend
- Partially completes #1205; and completes #1314 by CCPPizing compute_vdiff in diffusion_solver.F90
- Correctly update 'AST' field in physics buffer in RK stratiform CAM interface for snapshot consistency (bit-for-bit).
- Adds snapshot points for orographic form drag stress (Beljaars in CAM6+ and TMS in CAM5) for future CCPPization work.

Describe any changes made to build system:
- add holtslag_boville CCPP scheme
- add vertical_diffusion CCPP scheme

Describe any changes made to the namelist: N/A

List any changes to the defaults for the boundary datasets: N/A

Describe any substantial timing or memory changes: N/A

Code reviewed by:

List all files eliminated:
```
D       src/physics/cam/hb_diff.F90
  - moved to atmos_phys as CCPPized holstlag_boville scheme
D       src/physics/cam/vdiff_lu_solver.F90
  - moved to atmos_phys into to_be_ccppized (unchanged)
```

List all files added and what they do: N/A

List all existing files that have been modified, and describe the changes:
```
R053    src/physics/cam/diffusion_solver.F90    src/physics/cam/diffusion_solver_cam.F90
  - this is the "remnant" diffusion solver supporting WACCM with molecular diffusion.
  - partial refactor to avoid excessive code divergence with CCPPized diffusion_solver.
  - renamed to avoid name conflict with CCPPized diffusion_solver.

M       bld/configure
  - add holtslag_boville and diffusion_solver folders in atmos_phys/schemes to build.

M       src/physics/cam/vertical_diffusion.F90
  - use CCPPized subroutines for HB and vertical diffusion solver.

M       src/physics/cam/eddy_diff_cam.F90
  - use CCPPized vertical diffusion solver.

M       src/physics/cam/molec_diff.F90
  - dechunkize and use assumed dimensions for vd_lu_qdecomp.

M       src/physics/cam/clubb_intr.F90
  - use CCPPized HB boundary layer computation.

M       src/physics/cam/ref_pres.F90
M       src/utils/air_composition.F90
  - comment updates for clarification of units/purpose.

M       src/physics/cam/rk_stratiform_cam.F90
  - Correctly update AST pbuf field in RK stratiform CAM interface for snapshot consistency (bit-for-bit).

M       src/physics/cam7/physpkg.F90
M       src/physics/cam/physpkg.F90
M       bld/build-namelist
M       bld/namelist_files/namelist_definition.xml
  - move Beljaars, TMS drag call routines to physpkg level for snapshotting.
  - add orographic_form_drag_stress snapshot point.

M       src/control/cam_snapshot_common.F90
  - expand size of cam_in snapshot output maximum because it needs to fit all constituents.
```

Verified bit-for-bit with Derecho Intel prior to submitting

jimmielin committed:
Initial move of hb_diff code into atmos_phys.

jimmielin committed:
Dechunkize, remove pcols from compute_vdiff; general cleanup of compute_vdiff for future CCPPization

jimmielin committed:
Clean up compute_vdiff unused arguments; use dependencies

jimmielin committed:
Reword comment for readability

jimmielin committed:
Remove oop fieldlist in compute_vdiff and replace with logical array

jimmielin committed:
Add comments

jimmielin committed:
Merge branch 'cam_development' into hplin/hb_diff

jimmielin committed:
Fix wxie sms test diff due to wrong if and clause

jimmielin committed:
Move diffusion solver for non-WACCM to atmos_phys

jimmielin committed:
Fix bit-for-bitness in toa interface temperature interpolation for waccm-x vertical diff.

jimmielin committed:
Separate out vertical_diffusion_compute; tau_damp_rate

jimmielin committed:
Add use of new set of vdiff interstitials

jimmielin committed:
Merge branch 'cam_development' into hplin/hb_diff

jimmielin committed:
Update interface; fix build bugs

jimmielin committed:
Merge branch 'cam_development' into hplin/hb_diff

jimmielin committed:
Remove unused arguments

jimmielin committed:
Merge branch 'cam_development' into hplin/hb_diff

jimmielin committed:
Copy cloud fraction to ast (stratiform_cloud_area_fraction) for pbuf snapshot

jimmielin committed:
... wrong cloud field (use stratiform cloud not rh cloud)

jimmielin committed:
Merge branch 'hplin/rk_ast_pbuf_field' into hplin/hb_diff

jimmielin committed:
Merge branch 'cam_development' into hplin/hb_diff

jimmielin committed:
Zero out HB below where CLUBB is active in the free_atm_diff scheme.

jimmielin committed:
Output clubbtop pbuf field as real by converting from integer

nusbaume commented on src/physics/cam/vertical_diffusion.F90:
```diff
@@ -490,24 +482,8 @@ subroutine vertical_diffusion_init(pbuf2d)
      call addfld( 'UFLX'        , (/ 'ilev' /)  , 'A', 'W/m2'   , 'Zonal momentum flux'                                )
      call addfld( 'VFLX'        , (/ 'ilev' /)  , 'A', 'W/m2'   , 'Meridional momentm flux'                            )
```
I think (?) the units for these fields should be `N/m2`.  Otherwise they don't match what is in the CCPP-ized SIMA diagnostics.

jimmielin replied:
You're right, it seems like it was fixed in a later push https://github.com/jimmielin/CAM/commit/8b43f1d7b96a9c6181eea14c058f1e91f740d769

jimmielin committed:
Update for reviewer comments in atmos_phys.

jimmielin committed:
Use PBL diagnostics module in atmos_phys

jimmielin committed:
Change ztodt to dt in atmos_phys interfaces

jimmielin committed:
Merge branch 'cam_development' into hplin/hb_diff

jimmielin committed:
Fix build

jimmielin committed:
Merge in WACCM-X eddy_top handling to regular version.

jimmielin committed:
change cam7 flag to apply_nonwv_cflx for clarity

jimmielin committed:
Separate out TMS/Beljaars (orographic form drag) from vertical diffusion section in physpkg.

This will facilitate future CCPPization of the CAM7 vertical diffusion,
Beljaars and Turbulent Mountain Stress (TMS; CAM5) orographic form drag,
as otherwise their changes are embedded within the vertical_diffusion_tend and
the corresponding snapshot sections.

jimmielin committed:
Fix CAM7 snapshotting

jimmielin committed:
Merge branch 'cam_development' into hplin/hb_diff

jimmielin committed:
Merge branch 'hplin/separate_out_tmsblj_to_physpkg' into hplin/hb_diff

nusbaume commented on src/control/cam_snapshot_common.F90:
```diff
@@ -84,7 +84,7 @@ module cam_snapshot_common
 type (snapshot_type)    ::  cnst_snapshot(pcnst)
 type (snapshot_type)    ::  tend_snapshot(6)
-type (snapshot_type)    ::  cam_in_snapshot(30)
+type (snapshot_type)    ::  cam_in_snapshot(200)   ! needs to be bigger than pcnst because cam_in is split by constituent.
```
Given that `pcnst` could theoretically change given the CAM compset being used, should we make this dimension size dependent on `pcnst` itself?  For example:

```suggestion
type (snapshot_type)    ::  cam_in_snapshot(pcnst+31)   ! needs to be bigger than pcnst because cam_in is split by constituent.
```

Note that I just based the `31` number off of the original `30` plus one extra for padding, but if a different number works better there then I am happy with that instead.

nusbaume commented on src/control/cam_snapshot_common.F90:
```diff
@@ -1227,28 +1230,40 @@ subroutine cam_pbuf_snapshot_all_outfld(lchnk, file_num, pbuf)
+         ! integer-type fields.
+         if( trim(pbuf_snapshot(i)%ddt_string) == 'clubbtop') then
+            call pbuf_get_field(pbuf, pbuf_idx, tmpptr2d_int)
+            ! copy into real
+            allocate(tmpptr2d(size(tmpptr2d_int, 1), size(tmpptr2d_int, 2)))
+            tmpptr2d = real(tmpptr2d_int, r8)
+            call outfld(pbuf_snapshot(i)%standard_name, tmpptr2d, pcols, lchnk)
+            deallocate(tmpptr2d)
```
Do you need to allocate an entirely new variable here, or can you just do the following:

```suggestion
            ! convert into real
            call outfld(pbuf_snapshot(i)%standard_name, real(tmpptr2d_int, r8), pcols, lchnk)
```

jimmielin replied:
Thanks for the suggestion! I _think_ the `real(tmpptr2d_int, r8)` will also make a temporary and compile to the same thing. Do you mind if I just kept it as-is? I don't have a strong argument for it other than I would have to go back and remake the snapshot to retest it to be sure it works as intended, but if you have a strong opinion on cleaning this up I'm happy to make the change!

nusbaume replied:
If re-creating the snapshots and testing everything over again is a pain then I am totally fine to leave this as-is, as I do agree that the compiler will be doing the same thing under-the-hood.  Just note that for the (hopefully) longer-lived CAM-SIMA or atmospheric_physics repos I'll probably push harder against this.

Anyways, will go ahead and mark this comment as resolved now.  Thanks!

jimmielin replied:
Thanks @nusbaume! Yes I definitely agree with your change - my assumption is that this will be relatively short-lived. If I end up having to add more integer snapshot variables into this in the future I'll try to remember to make the suggested change.

nusbaume commented on src/physics/cam/diffusion_solver_cam.F90:
```diff
@@ -153,99 +105,82 @@ subroutine compute_vdiff( lchnk           ,
+    integer,  intent(in) :: pver
+    integer,  intent(in) :: pverp
+    integer,  intent(in) :: ncnst            ! # of constituents to diffuse. In eddy_diff, only wv. Others, pcnst.
+    real(r8), intent(in) :: ztodt            ! 2 delta-t [ s ]
```
I might remove the "2" here, as I believe that was only true for the Eulerian dycore (which is no longer in CAM):

```suggestion
    real(r8), intent(in) :: ztodt            ! delta-t [ s ]
```

nusbaume commented on src/physics/cam/diffusion_solver_cam.F90:
```diff
@@ -153,99 +105,82 @@ subroutine compute_vdiff( lchnk           ,
-    ! Optional Arguments !
-    ! ------------------ !
+    real(r8), intent(in)    :: cpairv(:,:)      ! Specific heat at constant pressure
+    real(r8), intent(in)    :: mbarv(:,:)       ! Composition dependent atmosphere mean mass
```
Science nit-pick, but is `mbarv` the mean mass (kg) or the mean molar mass (kg mol-1)?  Might be good to specify that in the comment here (just in case).

jimmielin replied:
Seems to be `kg mol-1` as it is initialized to `mwdry`; backing out where it is calculated in `air_composition::get_mbarv_1hd`

```fortran
             mm = tracer(idx, kdx, m_cnst) * factor(idx, kdx)
             mbarv_in(idx, kdx) = mbarv_in(idx, kdx) + thermodynamic_active_species_mwi(qdx) * mm
! ...
       mbarv_in(:,:) = 1.0_r8 / mbarv_in(:,:)
```

if `[tracer] = [kg kg(air)-1]` and `[mwi] = [1 / mw] = [mol kg-1]` it also works out that `[mbarv] = [kg mol-1]` so it is molar mass.

`air_composition.F90` says
```fortran
   ! get_mbarv: molecular weight of dry air
   public :: get_mbarv
...

   ! mbarv: composition dependent atmosphere mean mass
   real(r8), public, protected, allocatable :: mbarv(:,:,:)
```

so I'll also change the comment of `mbarv` here to "mean molar mass" to be consistent

nusbaume commented on src/physics/cam/eddy_diff_cam.F90:
```diff
@@ -200,18 +197,10 @@ subroutine eddy_diff_init(pbuf2d, ntop_eddy_in, nbot_eddy_in)
-       subname="vdiff_select")
-  call handle_errmsg(vdiff_select(fieldlist_wet,'v'), &
-       subname="vdiff_select")
+  ! Only diffuse constituent 1 and disable molecular diffusion
```
I might call out "water vapor" here specifically:

 ```suggestion
  ! Only diffuse water vapor (constituent 1) and disable molecular diffusion
```

nusbaume commented on src/physics/cam/eddy_diff_cam.F90:
```diff
@@ -464,7 +454,7 @@ subroutine compute_eddy_diff( pbuf, lchnk  ,
   real(r8), intent(in)    :: tauy(pcols)               ! Meridional wind stress at surface [ N/m2 ]
   real(r8), intent(in)    :: shflx(pcols)              ! Sensible heat flux at surface [ unit ? ]
-  real(r8), intent(in)    :: qflx(pcols)               ! Water vapor flux at surface [ unit ? ]
+  real(r8), intent(in)    :: qflx(pcols,1)             ! Water vapor flux at surface [ unit ? ]
```
If this is coming from `cflx` then I believe the units are `kg/m2/s`:

```suggestion
  real(r8), intent(in)    :: qflx(pcols,1)             ! Water vapor flux at surface [ kg/m2/s]
```

nusbaume commented on src/physics/cam/molec_diff.F90:
```diff
@@ -234,27 +234,26 @@ function vd_lu_qdecomp( &
     type(BoundaryType), intent(in) :: interface_boundary ! Boundary on grid edge.
     type(BoundaryType), intent(in) :: molec_boundary ! Boundary at edge of molec_diff region.
-    real(r8), intent(in)    :: tint(pcols,pver+1)    ! Interface temperature [ K ]
+    real(r8), intent(in)    :: tint(:,:)    ! Interface temperature [ K ], interfaces
```
Probably don't need to say "interface" twice in the comments here:

```suggestion
    real(r8), intent(in)    :: tint(:,:)    ! Air temperature [ K ], interfaces
```

nusbaume commented on src/physics/cam/molec_diff.F90:
```diff
@@ -234,27 +234,26 @@ function vd_lu_qdecomp( &
-    real(r8), intent(in)    :: t(pcols,pver)	    ! temperature
+    real(r8), intent(in)    :: t(:,:)	    ! temperature
     integer,  intent(in)    :: m 		    ! cnst index
+    real(r8), intent(in)    :: mbarv(:,:)  ! composition dependent atmosphere mean mass
```
I might update these comments slightly:

```suggestion
    real(r8), intent(in)    :: t(:,:)	    ! Air temperature [ K ]
    integer,  intent(in)    :: m 		    ! cnst index
    real(r8), intent(in)    :: mbarv(:,:)  ! composition dependent molar mass of dry air [ kg kmol-1]
```

nusbaume commented on src/physics/cam/vertical_diffusion.F90:
```diff
@@ -94,22 +90,32 @@ module vertical_diffusion
-type(vdiff_selector) :: fieldlist_dry                ! Logical switches for dry mixing ratio diffusion
-type(vdiff_selector) :: fieldlist_molec              ! Logical switches for molecular diffusion
+logical              :: do_diffusion_const_dry(pcnst)! Do vertical diffusion (as dry) for this constituent?
+logical              :: do_diffusion_const_wet(pcnst)! Do vertical diffusion (as wet) for this constituent?
```
I might specify what "dry" and "wet" specifically mean in this context:

```suggestion
logical              :: do_diffusion_const_dry(pcnst)! Do vertical diffusion (as a dry mixing ratio) for this constituent?
logical              :: do_diffusion_const_wet(pcnst)! Do vertical diffusion (as a wet mixing ratio) for this constituent?
```

nusbaume commented on src/physics/cam/vertical_diffusion.F90:
```diff
@@ -870,105 +906,180 @@ subroutine vertical_diffusion_tend( &
-  tautotx(:ncol) = cam_in%wsx(:ncol) + tautmsx(:ncol)
-  tautoty(:ncol) = cam_in%wsy(:ncol) + tautmsy(:ncol)
+  ! Add turbulent mountain stress to total surface stress for PBL scheme.
+  ! tautotx(:ncol) = cam_in%wsx(:ncol) + tautmsx(:ncol)
+  ! tautoty(:ncol) = cam_in%wsy(:ncol) + tautmsy(:ncol)
```
Remove commented-out code?

jimmielin replied:
Thanks, removed!

nusbaume commented on src/physics/cam/vertical_diffusion.F90:
```diff
@@ -870,105 +906,180 @@ subroutine vertical_diffusion_tend( &
+  q_wv_cflx(:) = 0._r8
+  !END REMOVECAM
+
+  if(eddy_scheme .eq. 'CLUBB_SGS') then
```
Nit-pick, but I might change the syntax here to better match the CAM standards:

```suggestion
  if(eddy_scheme == 'CLUBB_SGS') then
```

nusbaume commented on src/physics/cam/vertical_diffusion.F90:
```diff
@@ -1122,164 +1363,433 @@ subroutine vertical_diffusion_tend( &
+             errflg = errflg)
      else
-       cflux(:,2:) = cam_in%cflx(:,2:)
+        dse_top(:ncol) = 0._r8
```
I might add a comment here stating that setting `dse_top` to zero is just an artifact of how the routines behave, and is not supposed to be physically realistic.

jimmielin replied:
Thanks, added the same comment as in the atmos_phys code
```fortran
  ! Note: this is an artifact to the way the vertical diffusion calculations are done;
  ! it does not represent physical reality.
```

nusbaume commented on src/physics/cam/vertical_diffusion.F90:
```diff
@@ -281,59 +290,26 @@ subroutine vertical_diffusion_init(pbuf2d)
-        end do
-     end if !allocated
-  end if
+  call vertical_diffusion_sponge_layer_init( &
```
I think we should check the value of `errflg` at the end of each CCPP-ized scheme call in this module, and if non-zero call `endrun` with the `errmsg` string included in the final end run message.

We could also call `handle_errmsg` instead of `endrun` as well, as long as we make sure that `handle_errmsg` properly ends the model simulation instead of just printing the error to the log.

jimmielin replied:
Thanks, added! I ended up just using `endrun`

nusbaume reviewed (changes requested):
Thanks for getting the CCPP-ized version of HB working in CAM @jimmielin!  My only major request is to make sure that all of the error codes/flags/messages from the CCPP-ized subroutines are being checked after they are called, in order to make sure we are catching fatal errors at the right location.

Otherwise I just had some (hopefully) minor requests, and possibly some comments that are already out-of-date (and so could just be resolved immediately).  Thanks again for seeing this through!

jimmielin committed:
Address review comments; check errflg at end of CCPPized vertical diffusion calls

jimmielin committed:
Merge branch 'cam_development' into hplin/hb_diff

jimmielin committed:
Update src/physics/cam/diffusion_solver_cam.F90

Co-authored-by: Jesse Nusbaumer <nusbaume@ucar.edu>

jimmielin committed:
Update src/control/cam_snapshot_common.F90

Co-authored-by: Jesse Nusbaumer <nusbaume@ucar.edu>

jimmielin committed:
Update src/physics/cam/eddy_diff_cam.F90

Co-authored-by: Jesse Nusbaumer <nusbaume@ucar.edu>

jimmielin committed:
Update src/physics/cam/eddy_diff_cam.F90

Co-authored-by: Jesse Nusbaumer <nusbaume@ucar.edu>

jimmielin committed:
Update src/physics/cam/molec_diff.F90

Co-authored-by: Jesse Nusbaumer <nusbaume@ucar.edu>

jimmielin committed:
Update src/physics/cam/vertical_diffusion.F90

Co-authored-by: Jesse Nusbaumer <nusbaume@ucar.edu>

jimmielin committed:
Update src/physics/cam/molec_diff.F90

Co-authored-by: Jesse Nusbaumer <nusbaume@ucar.edu>

jimmielin committed:
Update src/physics/cam/vertical_diffusion.F90

Co-authored-by: Jesse Nusbaumer <nusbaume@ucar.edu>

jimmielin committed:
Address review comments (2)

jimmielin committed:
Fix build changes; do_beljaars logical move

jimmielin committed:
Fix TMS reset initialization in CAM7 (set to zero even if not used)

nusbaume reviewed (approved):
Everything looks good to me now.  Thanks again @jimmielin!

jimmielin committed:
Merge branch 'cam_development' into hplin/hb_diff

jimmielin committed:
Merge branch 'cam_development' into hplin/hb_diff

jimmielin committed:
Draft Changelog.

jimmielin committed:
Update to atmos_phys atmos_phys0_18_000

jimmielin committed:
Finalize ChangeLog for cam6_4_120 - complete CCPPization of HB and diffusion solver
